[v0.13] fail to export image when using rewrite-timestamp=true

Hello,

I've been running BuildKit `0.13.0-beta1` for a while as a `docker-container` driver to get caching ECR registry integration and timestamp rewriting for reproducibility. This worked well for a couple of months.

Today I tried upgrading BuildKit to the stable 0.13.0 and 0.13.1 releases with the same `docker-container` setup, and started encountering these errors:
```
...
#18 [10/10] COPY test.conf /etc/test.conf
#18 CACHED
#19 exporting to image
#19 exporting layers done
#19 rewriting layers with source-date-epoch 1 (1970-01-01 00:00:01 +0000 UTC)
------
 > exporting to image:
------
ERROR: failed to solve: content digest sha256:7b722c1070cdf5188f1f9e43b8413157f8dfb2b4fe84db3c03cb492379a42fcc: not found
```

For some reason jobs that use this cached layer fail to export it.
Once I delete the cache from ECR - 1st build passes, and subsequent builds fail.

Everything works again once I revert to using BuildKit `0.13.0-beta1`, with new and existing caches.

This is the command I use:
```
docker buildx build \
--build-arg SOURCE_DATE_EPOCH=1 \
--builder buildkit-6f9724f17cbb483a5f1e \
--provenance=false \
--sbom=false \
--cache-from type=registry,ref=0000000000.dkr.ecr.us-east-2.amazonaws.com/image-name:branch-name-cache \
--cache-from type=registry,ref=0000000000.dkr.ecr.us-east-2.amazonaws.com/image-name:main-cache \
--cache-to mode=max,image-manifest=true,oci-mediatypes=true,type=registry,ref=0000000000.dkr.ecr.us-east-2.amazonaws.com/image-name:branch-name \
--output type=registry,rewrite-timestamp=true \
--tag 0000000000.dkr.ecr.us-east-2.amazonaws.com/image-name:tmp.4521fd86 .
```

Edit: similar to [this issue](https://github.com/moby/buildkit/issues/4746), but I'm not doing `COPY --link`

Minimal repro:
```
sudo buildctl build --frontend dockerfile.v0 --local dockerfile=. --local context=. --no-cache --output type=image,name=localhost:5000/foo,push=true,rewrite-timestamp=true
```
(Needs `push=true` && `rewrite-timestamp=true`)


I also see an occasional panic in the daemon log
```
WARN[2024-05-15T02:44:50+09:00] rewrite-timestamp is specified, but no source-date-epoch was found  spanID=cbbd84754ca443c5 traceID=a347fa5861b3af65b3508699ef6f1797
panic: send on closed channel

goroutine 671 [running]:
github.com/containerd/containerd/remotes/docker.(*pushWriter).setPipe(...)
        /home/suda/gopath/src/github.com/moby/buildkit/vendor/github.com/containerd/containerd/remotes/docker/pusher.go:364
github.com/containerd/containerd/remotes/docker.dockerPusher.push.func1()
        /home/suda/gopath/src/github.com/moby/buildkit/vendor/github.com/containerd/containerd/remotes/docker/pusher.go:286 +0xcc
github.com/containerd/containerd/remotes/docker.(*request).do(0xc000d5b5f0, {0x295a620, 0xc000581410})
        /home/suda/gopath/src/github.com/moby/buildkit/vendor/github.com/containerd/containerd/remotes/docker/resolver.go:556 +0x162
github.com/containerd/containerd/remotes/docker.(*request).doWithRetries(0xc000d5b5f0, {0x295a620, 0xc000581410}, {0x0, 0x0, 0x0})
        /home/suda/gopath/src/github.com/moby/buildkit/vendor/github.com/containerd/containerd/remotes/docker/resolver.go:600 +0x47
github.com/containerd/containerd/remotes/docker.dockerPusher.push.func2()
        /home/suda/gopath/src/github.com/moby/buildkit/vendor/github.com/containerd/containerd/remotes/docker/pusher.go:292 +0x57
created by github.com/containerd/containerd/remotes/docker.dockerPusher.push in goroutine 787
        /home/suda/gopath/src/github.com/moby/buildkit/vendor/github.com/containerd/containerd/remotes/docker/pusher.go:291 +0x2150
```

--- a/exporter/containerimage/export.go
+++ b/exporter/containerimage/export.go
@@ -328,18 +328,7 @@ func (e *imageExporterInstance) Export(ctx context.Context, src *exporter.Source
 				}
 			}
 			if e.push {
-				if opts.RewriteTimestamp {
-					annotations := map[digest.Digest]map[string]string{}
-					addAnnotations(annotations, *desc)
-					// e.pushImage cannot be used because src ref does not point to the rewritten image
-					//
-					// TODO: change e.pushImage so that it takes Result[Remote] as parameter.
-					// https://github.com/moby/buildkit/pull/4057#discussion_r1324106088
-					err = push.Push(ctx, e.opt.SessionManager, sessionID, e.opt.ImageWriter.opt.ContentStore, e.opt.ImageWriter.ContentStore(),
-						desc.Digest, targetName, e.insecure, e.opt.RegistryHosts, e.pushByDigest, annotations)
-				} else {
-					err = e.pushImage(ctx, src, sessionID, targetName, desc.Digest)
-				}
+				err = e.pushImage(ctx, src, sessionID, targetName, desc.Digest)
 				if err != nil {
 					return nil, nil, errors.Wrapf(err, "failed to push %v", targetName)
 				}

--- a/frontend/dockerfile/dockerfile_test.go
+++ b/frontend/dockerfile/dockerfile_test.go
@@ -6882,6 +6882,7 @@ func testReproSourceDateEpoch(t *testing.T, sb integration.Sandbox) {
 		dockerfile     string
 		files          []fstest.Applier
 		expectedDigest string
+		noCacheExport  bool
 	}
 	testCases := []testCase{
 		{
@@ -6908,6 +6909,14 @@ COPY --link foo foo
 `,
 			files:          []fstest.Applier{fstest.CreateFile("foo", []byte("foo"), 0600)},
 			expectedDigest: "sha256:9f75e4bdbf3d825acb36bb603ddef4a25742afb8ccb674763ffc611ae047d8a6",
+		},
+		{
+			// https://github.com/moby/buildkit/issues/4793
+			name: "NoAdditionalLayer",
+			dockerfile: `FROM amd64/debian:bullseye-20230109-slim
+`,
+			expectedDigest: "sha256:eeba8ef81dec46359d099c5d674009da54e088fa8f29945d4d7fb3a7a88c450e",
+			noCacheExport:  true, // "skipping cache export for empty result"
 		},
 	}
 
@@ -6968,7 +6977,10 @@ COPY --link foo foo
 			require.NoError(t, err)
 
 			desc, manifest, img := readImage(t, ctx, target)
-			_, cacheManifest, _ := readImage(t, ctx, target+"-cache")
+			var cacheManifest ocispecs.Manifest
+			if !tc.noCacheExport {
+				_, cacheManifest, _ = readImage(t, ctx, target+"-cache")
+			}
 			t.Log("The digest may change depending on the BuildKit version, the snapshotter configuration, etc.")
 			require.Equal(t, tc.expectedDigest, desc.Digest.String())
 
@@ -6986,9 +6998,11 @@ COPY --link foo foo
 					require.Equal(t, fmt.Sprintf("%d", tm.Unix()), l.Annotations["buildkit/rewritten-timestamp"])
 				}
 			}
-			// Cache layers must *not* have rewritten-timestamp
-			for _, l := range cacheManifest.Layers {
-				require.Empty(t, l.Annotations["buildkit/rewritten-timestamp"])
+			if !tc.noCacheExport {
+				// Cache layers must *not* have rewritten-timestamp
+				for _, l := range cacheManifest.Layers {
+					require.Empty(t, l.Annotations["buildkit/rewritten-timestamp"])
+				}
 			}
 
 			// Build again, after pruning the base image layer cache.

